interfaces/machine_parameters/ring.py

```diff
@@ -0,0 +1,588 @@
+                                        self.f_rev)
+        parameters['t_rev'] = np.interp(cycle_moments, self.cycle_time,
+                                        self.t_rev)
+        parameters['omega_rev'] = np.interp(cycle_moments, self.cycle_time,
+                                            self.omega_rev)
+        parameters['eta_0'] = np.interp(cycle_moments, self.cycle_time,
+                                        self.eta_0[0])
+        if len(self.cycle_time) == len(self.delta_E[0]):
+            parameters['delta_E'] = np.interp(cycle_moments,
+                                              self.cycle_time,
+                                              self.delta_E[0])
+        else:
+            parameters['delta_E'] = np.interp(cycle_moments,
+                                              self.cycle_time[:-1],
+                                              self.delta_E[0])
+        parameters['charge'] = self.Particle.charge
+        parameters['cycle_time'] = cycle_moments
+
+        return parameters
+
+    def _no_parameters_at_turn(self, turn):
+        raise RuntimeError("parameters_at_turn only available if " +
+                           "store_turns = True at object declaration")
+
+    def _parameters_at_turn(self, turn):
+
+        try:
+            sample = np.where(self.use_turns == turn)[0][0]
+        except IndexError:
+            raise excpt.InputError("turn " + str(turn) + " has not been "
+                                   + "stored for the specified interpolation")
+        else:
+            return self.parameters_at_sample(sample)
+
+    def parameters_at_sample(self, sample):
+
+        parameters = {}
+        parameters['momentum'] = self.momentum[0, sample]
+        parameters['beta'] = self.beta[0, sample]
+        parameters['gamma'] = self.gamma[0, sample]
+        parameters['energy'] = self.energy[0, sample]
+        parameters['kin_energy'] = self.kin_energy[0, sample]
+        parameters['f_rev'] = self.f_rev[sample]
+        parameters['t_rev'] = self.t_rev[sample]
+        parameters['omega_rev'] = self.omega_rev[sample]
+        parameters['eta_0'] = self.eta_0[0, sample]
+        try:
+            parameters['delta_E'] = self.delta_E[0, sample]
+        except IndexError:
+            parameters['delta_E'] = self.delta_E[0, sample - 1]
+        parameters['charge'] = self.Particle.charge
+        parameters['cycle_time'] = self.cycle_time[sample]
+
+        return parameters
+
+    def _recalc_delta_E(self):
+        """
+        Function to recalculate delta_E.
+        If interpolation is not done on a turn-by-turn basis the delta_E will
+        not be correct.  This function recalculates it to give the correct
+        value for each turn.
+        """
+
+        for section in range(self.n_sections):
+            derivative = calc.deriv_cubic(self.cycle_time,
+                                          self.energy[section])[1]
+            self.delta_E[section] = derivative * self.t_rev
```

Comment:
This function does odd things when there's few points and a transition from flat bottom to ramp.  I can't find one that resolves the ramp well and also does the flat bottom well.  Know of anything better?

Reply:
I'd avoid cubic_spline here, especially for cases where Bdot is defined as piecewise linear functions, maybe worth looking into the linear functions instead but not sure if gradient or a modified version of diff (we would have the issue that the output is one element short) is best
.

Reply:
Yeah, I tried gradient and diff, but they give different problems elsewhere.  This gave the best results on the ramp, but then the flat bottom went wrong.  Never mind, I'll worry about it another time.